content/posts/2024-11-05-flang-new.md

```diff
@@ -0,0 +1,898 @@
+all the threads available to the program.
+
+Clang has [supported OpenMP](https://blog.llvm.org/2015/05/openmp-support_22.html)
+for many years now. The equivalent C++ code is:
+```C++
+void simple(int n, float *a, float *b)
+{
+    int i;
+
+    #pragma omp parallel for
+    for (i=1; i<n; i++)
+        b[i] = (a[i] + a[i-1]) / 2.0;
+}
+```
+([Compiler Explorer](https://godbolt.org/z/Yh9jb8rKe))
+
+In C++'s case, the directive is in the form of a `#pragma`, and attached
+to the `for` loop.
+
+LLVM IR does not know anything about OpenMP specifically, so Clang does all the
+work of converting the intent of the directives into LLVM IR. The output from
+Clang looks like this:
+
+```mlir
+define dso_local void @simple(int, float*, float*)
+  (i32 noundef %n, ptr noundef %a, ptr noundef %b) <...> {
+entry:
+<...>
+  call void (<...>) @__kmpc_fork_call(@simple <...> (.omp_outlined) <...>)
+  ret void
+}
+
+define internal void @simple(int, float*, float*) (.omp_outlined)
+  (ptr <...> %.global_tid.,
+   ptr <...> %.bound_tid.,
+   ptr <...> %n,
+   ptr <...> %b,
+   ptr <...> %a) {
+entry:
+<...>
+  call void @__kmpc_for_static_init_4(<...>)
+<...>
+omp.inner.for.body.i:
+<...>
+omp.loop.exit.i:
+  call void @__kmpc_for_static_fini(<...>)
+<...>
+  ret void
+}
+```
+(output edited for readability)
+
+The body of `simple` no longer does all the work. Instead it uses
+`__kmpc_fork_call` to tell the OpenMP
+[runtime library](https://github.com/llvm/llvm-project/tree/main/openmp)
+to run another function, `simple (.omp_outlined)` to do the work.
+
+This second function is referred to as a "micro task". The runtime library
+splits the work across many instances of the micro task and each time
+the micro task function is called, it gets a different slice of the work.
+
+The number of instances is only known at runtime, and can be controlled with
+settings such as [`OMP_NUM_THREADS`](https://www.openmp.org/spec-html/5.0/openmpse50.html).
+
+This function's LLVM IR includes labels like `omp.loop.exit.i` but these are
+not specific to OpenMP. They are just normal LLVM IR labels whose name includes
+`omp`.
+
+## Sharing Clang's OpenMP Knowledge
+
+In April 2019 LLVM Flang ("F18" at the time) was approved to join the LLVM
+Project. A month later, it was proposed that the soon-to-be LLVM Flang should
+leverage Clang's knowledge of OpenMP.
+
+> This is an RFC for the design of the OpenMP front-ends under the LLVM
+> umbrella. It is necessary to talk about this now as Flang (aka. F18) is
+> maturing at a very promising rate and about to become a sub-project next
+> to Clang.
+>
+> TLDR;
+> Keep AST nodes and Sema separated but unify LLVM-IR generation for
+> OpenMP constructs based on the (almost) identical OpenMP directive
+> level.
+- "[RFC] Proposed interplay of Clang & Flang & LLVM wrt. OpenMP",
+  Johannes Doerfert, May 2019 (from a copy provided to me, no copies appear
+  to exist online at this time)
```

Comment:
Here's my small contribution: the RFC is still available. You can include it for completeness: [RFC: Proposed Interplay of Clang, Flang, LLVM wrt OpenMP](https://discourse.llvm.org/t/rfc-proposed-interplay-of-clang-flang-llvm-wrt-openmp-flang-dev/51905).

Reply:
Done.